components/server/src/workspace/workspace-starter.ts

```diff
@@ -690,122 +689,149 @@ export class WorkspaceStarter {
-        }
+            if (forcePVC === true) {
+                featureFlags = featureFlags.concat(["persistent_volume_claim"]);
+            }
 
-        if (!!featureFlags) {
-            // only set feature flags if there actually are any. Otherwise we waste the
-            // few bytes of JSON in the database for no good reason.
-            configuration.featureFlags = featureFlags;
-        }
+            if (!!featureFlags) {
+                // only set feature flags if there actually are any. Otherwise we waste the
+                // few bytes of JSON in the database for no good reason.
+                configuration.featureFlags = featureFlags;
+            }
 
-        const usageAttributionId = await this.userService.getWorkspaceUsageAttributionId(user, workspace.projectId);
-
-        const now = new Date().toISOString();
-        const instance: WorkspaceInstance = {
-            id: uuidv4(),
-            workspaceId: workspace.id,
-            creationTime: now,
-            ideUrl: "", // Initially empty, filled during starting process
-            region: this.config.installationShortname, // Shortname set to bridge can cleanup workspaces stuck preparing
-            workspaceImage: "", // Initially empty, filled during starting process
-            status: {
-                version: 0,
-                conditions: {},
-                phase: "preparing",
-            },
-            configuration,
-            usageAttributionId,
-        };
-        if (WithReferrerContext.is(workspace.context)) {
-            this.analytics.track({
-                userId: user.id,
-                event: "ide_referrer",
-                properties: {
-                    workspaceId: workspace.id,
-                    instanceId: instance.id,
-                    referrer: workspace.context.referrer,
-                    referrerIde: workspace.context.referrerIde,
-                },
+            const usageAttributionId = await this.userService.getWorkspaceUsageAttributionId(user, workspace.projectId);
+
+            let workspaceClass = "";
+            let classesEnabled = await getExperimentsClientForBackend().getValueAsync("workspace_classes", false, {
+                user: user,
             });
+            if (classesEnabled) {
+                // this is a workspace that was started before workspace classes
+                // set the workspace class based on if the user "has more resources"
+                if (!workspace.workspaceClass) {
+                    if (await this.userService.userGetsMoreResources(user)) {
+                        workspaceClass = this.config.workspaceClasses.defaultMoreResources;
+                    } else {
+                        workspaceClass = this.config.workspaceClasses.default;
+                    }
+
+                    workspace.workspaceClass = workspaceClass;
+                    this.workspaceDb.trace({ span }).store(workspace);
```

Comment:
nit: there should be a .updatePartial() method which you can call with just `updatePartial({workspaceClass})`. This is slightly safer, as you never now who else does what changes to the `workspace` shape at hand.

Reply:
Also: please add a trailing `.catch(..)` - or `await` the call.